packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts

```diff
@@ -37,174 +45,238 @@ export const GridRootStyles = styled('div', {
+    : t.palette.background.default;
+
+  const pinnedBackground =
+    t.palette.mode === 'dark'
+      ? lighten(
+          t.vars ? t.vars.palette.background.defaultChannel : t.palette.background.default,
+          0.1,
+        )
+      : darken(
+          t.vars ? t.vars.palette.background.defaultChannel : t.palette.background.default,
+          0.05,
+        );
+
+  const overlayBackground = t.vars
+    ? `rgba(${t.vars.palette.background.defaultChannel} / ${t.vars.palette.action.disabledOpacity})`
+    : alpha(t.palette.background.default, t.palette.action.disabledOpacity);
+
+  const hoverOpacity = (t.vars || t).palette.action.hoverOpacity;
+  const hoverBackground = (t.vars || t).palette.action.hover;
+
+  const selectedOpacity = (t.vars || t).palette.action.selectedOpacity;
+  const selectedBackground = t.vars
+    ? `rgba(${t.vars.palette.primary.mainChannel} / ${selectedOpacity})`
+    : alpha(t.palette.primary.main, selectedOpacity);
+
+  const selectedHoverBackground = t.vars
+    ? `rgba(${t.vars.palette.primary.mainChannel} / calc(
+                ${t.vars.palette.action.selectedOpacity} + 
+                ${t.vars.palette.action.hoverOpacity}
+              ))`
+    : alpha(
+        t.palette.primary.main,
+        t.palette.action.selectedOpacity + t.palette.action.hoverOpacity,
+      );
+
+  const pinnedHoverBackground = blend(pinnedBackground, hoverBackground, hoverOpacity);
+  const pinnedSelectedBackground = blend(pinnedBackground, selectedBackground, selectedOpacity);
+  const pinnedSelectedHoverBackground = blend(
+    pinnedSelectedBackground,
+    hoverBackground,
+    hoverOpacity,
+  );
+
+  const selectedStyles = {
+    backgroundColor: selectedBackground,
+    '&:hover': {
+      backgroundColor: selectedHoverBackground,
+      // Reset on touch devices, it doesn't add specificity
+      '@media (hover: none)': {
+        backgroundColor: selectedBackground,
+      },
+    },
+  };
 
   const gridStyle: CSSInterpolation = {
     '--unstable_DataGrid-radius': typeof radius === 'number' ? `${radius}px` : radius,
-    '--unstable_DataGrid-headWeight': theme.typography.fontWeightMedium,
-    '--unstable_DataGrid-overlayBackground': theme.vars
-      ? `rgba(${theme.vars.palette.background.defaultChannel} / ${theme.vars.palette.action.disabledOpacity})`
-      : alpha(theme.palette.background.default, theme.palette.action.disabledOpacity),
+    '--unstable_DataGrid-headWeight': t.typography.fontWeightMedium,
+    '--unstable_DataGrid-overlayBackground': overlayBackground,
+
+    '--DataGrid-containerBackground': containerBackground,
+    '--DataGrid-pinnedBackground': pinnedBackground,
+    '--DataGrid-rowBorderColor': borderColor,
+
     '--DataGrid-cellOffsetMultiplier': 2,
+    '--DataGrid-width': '0px',
+    '--DataGrid-hasScrollX': '0',
+    '--DataGrid-hasScrollY': '0',
+    '--DataGrid-offsetTop': '0px',
+    '--DataGrid-offsetLeft': '0px',
+    '--DataGrid-scrollbarSize': '10px',
```

Comment:
I looked into the point (1) from https://github.com/mui/mui-x/pull/10059#discussion_r1417075038 and managed to fix the automatic scrollbar on MacOS with the following diff:

```diff
diff --git a/packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts b/packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts
index 836dd6392..11fd2e6e9 100644
--- a/packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts
+++ b/packages/grid/x-data-grid/src/components/containers/GridRootStyles.ts
@@ -198,7 +198,6 @@ export const GridRootStyles = styled('div', {
     '--DataGrid-hasScrollY': '0',
     '--DataGrid-offsetTop': '0px',
     '--DataGrid-offsetLeft': '0px',
-    '--DataGrid-scrollbarSize': '10px',
     '--DataGrid-rowWidth': '0px',
     '--DataGrid-columnsTotalWidth': '0px',
     '--DataGrid-leftPinnedWidth': '0px',
diff --git a/packages/grid/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx b/packages/grid/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx
index 87a2b7779..af94c326c 100644
--- a/packages/grid/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx
+++ b/packages/grid/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx
@@ -37,13 +37,13 @@ const Scrollbar = styled('div')({
 });
 
 const ScrollbarVertical = styled(Scrollbar)({
-  width: 'var(--DataGrid-scrollbarSize)',
+  width: 'var(--DataGrid-scrollbarSize, 16px)',
   height:
-    'calc(var(--DataGrid-hasScrollY) * (100% - var(--DataGrid-topContainerHeight) - var(--DataGrid-bottomContainerHeight) - var(--DataGrid-hasScrollX) * var(--DataGrid-scrollbarSize)))',
+    'calc(var(--DataGrid-hasScrollY) * (100% - var(--DataGrid-topContainerHeight) - var(--DataGrid-bottomContainerHeight) - var(--DataGrid-hasScrollX) * var(--DataGrid-scrollbarSize, 16px)))',
   overflowY: 'auto',
   overflowX: 'hidden',
   '& > div': {
-    width: 'var(--DataGrid-scrollbarSize)',
+    width: 'var(--DataGrid-scrollbarSize, 16px)',
   },
   top: 'var(--DataGrid-topContainerHeight)',
   right: '0px',
@@ -51,11 +51,11 @@ const ScrollbarVertical = styled(Scrollbar)({
 
 const ScrollbarHorizontal = styled(Scrollbar)({
   width: '100%',
-  height: 'var(--DataGrid-scrollbarSize)',
+  height: 'var(--DataGrid-scrollbarSize, 16px)',
   overflowY: 'hidden',
   overflowX: 'auto',
   '& > div': {
-    height: 'var(--DataGrid-scrollbarSize)',
+    height: 'var(--DataGrid-scrollbarSize, 16px)',
   },
   bottom: '0px',
 });
diff --git a/packages/grid/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts b/packages/grid/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts
index a16662db6..395f4b5db 100644
--- a/packages/grid/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts
+++ b/packages/grid/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts
@@ -307,7 +307,9 @@ export function useGridDimensions(
     set('--DataGrid-width', `${dimensions.viewportOuterSize.width}px`);
     set('--DataGrid-hasScrollX', `${Number(dimensions.hasScrollX)}`);
     set('--DataGrid-hasScrollY', `${Number(dimensions.hasScrollY)}`);
-    set('--DataGrid-scrollbarSize', `${dimensions.scrollbarSize}px`);
+    if (dimensions.scrollbarSize !== 0) {
+      set('--DataGrid-scrollbarSize', `${dimensions.scrollbarSize}px`);
+    }
     set('--DataGrid-rowWidth', `${dimensions.rowWidth}px`);
     set('--DataGrid-columnsTotalWidth', `${dimensions.columnsTotalWidth}px`);
     set('--DataGrid-leftPinnedWidth', `${dimensions.leftPinnedWidth}px`);
```

Chromium:

https://github.com/mui/mui-x/assets/13808724/9a2f5794-fa3e-46e5-b611-b46e36ca0c30

Safari:

https://github.com/mui/mui-x/assets/13808724/0632de40-c8fc-4ea8-bdd8-43f8493e4b36

Scrollbars do not show up on iOS though - I will investigate this too.

Reply:
Solved, but this needed to go on the scrollbar container directly. For the rest of the grid, the scrollbar size needs to be `0px` otherwise stuff isn't positionned properly.

I've set it to an absolute value of 14px but I don't like that solution, but there's no way to figure out the scrollbar size for platforms where the scrollbar doesn't affect layout like macOS or gnome. In particular if the user is zoomed to a different level than 100%, the scrollbar "pixel" size could be much smaller or larger.

Reply:
Did you have time to test on iOS by any chance? I don't have any apple device, so I can't test it easily.

Reply:
Unfortunately, the scrollbars aren't visible on iOS:


https://github.com/mui/mui-x/assets/13808724/267bfab9-3b7d-444f-b4b6-daf9954e8175

We can fix this later though

Reply:
> Unfortunately, the scrollbars aren't visible on iOS:

I confirm this to be an issue, there's also another interesting observation on iOS.

When I scroll fast on this demo, there appears to be a lot of white space, but only when checkbox column is in view, if I scroll to the right, it appears to be quite smooth. Checkbox rendering seems to be expensive.

https://github.com/mui/mui-x/assets/12609561/21186c12-6ead-44a6-97ee-b68e8753acea

We could probably create a followup issue for these iOS related issues. 
CC @romgrk